api/src/main/java/org/projectnessie/tools/catalog/migration/api/CatalogMigrator.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright (C) 2023 Dremio
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.tools.catalog.migration.api;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import java.util.List;
+import java.util.Objects;
+import java.util.function.Predicate;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import org.apache.iceberg.BaseTable;
+import org.apache.iceberg.TableOperations;
+import org.apache.iceberg.catalog.Catalog;
+import org.apache.iceberg.catalog.Namespace;
+import org.apache.iceberg.catalog.SupportsNamespaces;
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.iceberg.hadoop.HadoopCatalog;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CatalogMigrator {
+  private static final Logger LOG = LoggerFactory.getLogger(CatalogMigrator.class);
+
+  public static final String FAILED_IDENTIFIERS_FILE = "failed_identifiers.txt";
+  public static final String FAILED_TO_DELETE_AT_SOURCE_FILE = "failed_to_delete_at_source.txt";
+  public static final String DRY_RUN_FILE = "dry_run_identifiers.txt";
+
+  private final Catalog sourceCatalog;
+  private final Catalog targetCatalog;
+  private final boolean deleteEntriesFromSourceCatalog;
+
+  public CatalogMigrator(CatalogMigratorParams catalogMigratorParams) {
+    this.sourceCatalog = catalogMigratorParams.sourceCatalog();
+    this.targetCatalog = catalogMigratorParams.targetCatalog();
+    this.deleteEntriesFromSourceCatalog = catalogMigratorParams.deleteEntriesFromSourceCatalog();
+  }
+
+  /**
+   * Get the table identifiers which matches the regular expression pattern input from all the
+   * namespaces.
+   *
+   * @param identifierRegex regular expression pattern. If null, fetches all the table identifiers
+   *     from all the namespaces.
+   * @return List of table identifiers.
+   */
+  public List<TableIdentifier> getMatchingTableIdentifiers(String identifierRegex) {
+    LOG.info("Collecting all the namespaces from source catalog...");
+    // fetch all the table identifiers from all the namespaces.
+    List<Namespace> namespaces =
+        (sourceCatalog instanceof SupportsNamespaces)
+            ? ((SupportsNamespaces) sourceCatalog).listNamespaces()
+            : ImmutableList.of(Namespace.empty());
```

Comment:
There's is a bug for catalogs that do not support namespaces - see why unit tests near the code make sense??